File: java/tools/src/main/java/io/github/ajoz/workshop/fp/java/tools/flow/PeekFlow.java
package io.github.ajoz.workshop.fp.java.tools.flow;

import io.github.ajoz.workshop.fp.java.tools.Consumer1;
import io.github.ajoz.workshop.fp.java.tools.control.Try;

final class PeekFlow<A> implements Flow<A> {
    private final Flow<A> upstream;
    private final Consumer1<? super A> action;

    PeekFlow(final Flow<A> upstream,
             final Consumer1<? super A> action) {
        this.upstream = upstream;
        this.action = action;
    }

    @Override
    public Try<A> next() {
        // Try implementation has ifSucces already so it's not necessary
        // to implement it manually
        return upstream.next().ifSuccess(action);
    }
}

File: src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/tools/flow/PeekFlow.kt
@file:Suppress("unused")

package io.github.ajoz.workshop.fp.kotlin.tools.flow

import io.github.ajoz.workshop.fp.kotlin.tools.Try

/**
 *
 */
fun <A> Flow<A>.peek(effect: (A) -> Unit): Flow<A> {
    class PeekFlow<A>(val upstream: Flow<A>,
                      val action: (A) -> Unit) : Flow<A> {

        override fun next(): Try<A> =
                upstream.next().ifSuccess(action)
    }

    return PeekFlow(this, effect)
}
